cql3/statements/modification_statement.cc

```diff
@@ -275,21 +273,46 @@ modification_statement::do_execute(query_processor& qp, service::query_state& qs
         return execute_with_condition(qp, qs, options);
     }
 
-    return execute_without_condition(qp, qs, options).then([] (coordinator_result<> res) {
+    json_cache_opt json_cache = maybe_prepare_json_cache(options);
+    std::vector<dht::partition_range> keys = build_partition_keys(options, json_cache);
+
+    return execute_without_condition(qp, qs, options, json_cache, keys).then([this, &keys] (coordinator_result<> res) {
         if (!res) {
             return make_ready_future<::shared_ptr<cql_transport::messages::result_message>>(
                     seastar::make_shared<cql_transport::messages::result_message::exception>(std::move(res).assume_error()));
         }
-        return make_ready_future<::shared_ptr<cql_transport::messages::result_message>>(
+
+        if (keys.size() != 1) {
+            return make_ready_future<::shared_ptr<cql_transport::messages::result_message>>(
                 ::shared_ptr<cql_transport::messages::result_message>{});
+        }
+
+        auto result = seastar::make_shared<cql_transport::messages::result_message::void_message>();
+
+        auto&& table = s->table();
+        if (table.uses_tablets()) {
+            auto token = keys[0].start()->value().token();
+            auto shard = service::storage_proxy::cas_shard(*s, token);
+
+            auto erm = table.get_effective_replication_map();
+            auto tablet_replicas = erm->get_replicas_for_reading(token);
+            auto token_range = erm->get_token_range(token);
+
+            std::cout << "modification2\n";
+            if (!tablet_replicas.empty() && (!locator::contains(tablet_replicas, erm->get_token_metadata().get_my_id()) || this_shard_id() != shard)) {
```

Comment:
This logic is duplicated between select_statement and modification_statement. If we had erm->check_locality(), it would be contained there. It can return optional<tablet_routing_info>

Reply:
It's a new struct which will hold tablet replicas and token range, in case we need to return the info to the driver.

Reply:
Okay, but how this is supposed to make performance better? We still need to use `get_replicas_for_reading(token)` and check all the conditions, we just have it in separate function. Maybe I misunderstood something

Reply:
Yes, but you can inline it and avoid the copy and second lookup for obtaining a token range.